pkg/tls/tls.go

```diff
@@ -0,0 +1,281 @@
+// successfully fetched, for the context had been canceled.
+func (p *Provider) mustFetchCertificate(ctx context.Context) {
+	// Time to attempt to fetch a new certificate if the last failed.
+	ticker := time.NewTicker(time.Second * 20)
+	defer ticker.Stop()
+
+	for {
+		// Fetch a new serving certificate, signed by cert-manager.
+		if err := p.fetchCertificate(ctx); err != nil {
+			p.log.Errorf("failed to fetch new serving certificate: %s, retrying", err)
+
+			// Cancel if the context has been canceled. Retry after tick.
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				continue
+			}
+		}
+
+		p.log.Info("fetched new serving certificate")
+
+		return
+	}
+}
+
+// TLSConfig should be used by consumers of the provider to get a TLS config
+// which will have the signed certificate and private key appropriately renewed
+func (p *Provider) TLSConfig() (*tls.Config, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	if p.tlsConfig == nil {
+		return nil, errors.New("provider not configured, TLS config not ready")
+	}
+
+	return &tls.Config{
+		GetConfigForClient: p.getConfigForClient,
+		ClientAuth:         tls.RequireAndVerifyClientCert,
+	}, nil
+}
+
+// hetConfigForClient will return a TLS config based upon the current signed
+// certificate and private key the provider holds.
+func (p *Provider) getConfigForClient(_ *tls.ClientHelloInfo) (*tls.Config, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return p.tlsConfig, nil
+}
+
+// RootCA returns the configured CA certificate
+func (p *Provider) RootCA() []byte {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return p.rootCA
+}
+
+// fetchCertificate will attempt to fetch a new signed certificate with a new
+// private key for serving. This will then be stored as the latest TLS config
+// for this provider to be fetched by new client connections. If this process
+// fails, returns error.
+func (p *Provider) fetchCertificate(ctx context.Context) error {
+	opts := pkiutil.CertOptions{
+		// TODO: allow configurable namespace and service
+		Host:       "cert-manager-istio-agent.cert-manager.svc",
+		IsServer:   true,
+		TTL:        p.servingCertificateTTL,
+		RSAKeySize: 2048,
+	}
+
+	// Generate new CSR and private key for serving
+	csr, pk, err := pkiutil.GenCSR(opts)
+	if err != nil {
+		return fmt.Errorf("failed to generate serving private key and CSR: %s", err)
+	}
+
+	// Build the CertificateRequest for a serving certificate for this agent
+	// using the configured issuer.
+	cr := &cmapi.CertificateRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "cert-manager-istio-agent-",
+			Annotations: map[string]string{
+				"istio.cert-manager.io/identities": "cert-manager-istio-agent",
+			},
+		},
+		Spec: cmapi.CertificateRequestSpec{
+			Duration: &metav1.Duration{
+				Duration: p.servingCertificateTTL,
+			},
+			IsCA:      false,
+			Request:   csr,
+			Usages:    []cmapi.KeyUsage{cmapi.UsageServerAuth},
```

Comment:
I do wonder here, is this enough? this disables all (classic) KeyUsages?

Reply:
This is enough for my testing. This is just for serving the cert service for istio proxies (envoy). This passes e2e tests with 1.7